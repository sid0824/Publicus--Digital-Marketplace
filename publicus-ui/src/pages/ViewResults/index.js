import React from 'react';
import { API } from "aws-amplify";
import { parse } from "query-string";
import {Button, Collapse, message, Icon} from "antd";
import moment from "moment";
import { ApiName } from "../../core/Config";
import { Parser } from 'json2csv';
import download from 'js-file-download';

const { Panel } = Collapse;

/**
 * Page used to view results for a form owned or purchased by a user
 */
class ViewResults extends React.Component {

  state = {
    results: null // - The formatted results of the form
  };

  /**
   * Fetches form config and results data and formats it into answer string, prompt string and submission time
   * @returns {Promise<void>}
   */
  componentDidMount = async () => {
    const { location } = this.props;
    const parsed = parse(location.search);
    const { surveyId } = parsed;
    const promises = await Promise.all([
      API.get(ApiName, `/results/${surveyId}`, {}),
      API.get(ApiName, `/survey/${surveyId}`, {})
    ]);
    const results = promises[0];
    const survey = promises[1];
    const surveyConfig = JSON.parse(unescape(survey.form));
    this.setState({
      results: results.map(result => {
        const resultsJson = JSON.parse(unescape(result.results));
        return {
          submissionTime: result.submissionTime,
          data: Object.keys(resultsJson).map(id => {
            const answer = resultsJson[id];
            const prompt = surveyConfig[id].prompt;
            const type = surveyConfig[id].type;
            return {
              answer,
              prompt,
              type
            }
          })
        }
      })
    })
  };

  /**
   * Exports the existing data as a CSV file to the user
   */
  export = () => {
    const results = JSON.parse(JSON.stringify(this.state.results));
    if (results.length) {
      const formattedResults = results.map(result => {
        const formatted = {};
        result.data.forEach(answer => {
          formatted[answer.prompt] = answer.answer
        });
        return {
          submissionTime: result.submissionTime,
          ...formatted
        }
      });
      const parser = new Parser(Object.keys(formattedResults));
      const csv = parser.parse(formattedResults);
      download(csv, 'resultsData.csv');
    } else {
      message.error('No results to download')
    }
  };

  /**
   * Formats a response based on the variable type of the question
   * @param question - to format
   */
  formatAnswer = (question) => {
    if (question.type === 'switch') {
      const value = !!question.answer;
      return value.toString();
    }
    if (question.type === 'timePicker') {
      return question.answer
        ? moment(question.answer).format('h:mm:ss a')
        : null
    }
    if (question.type === 'datePicker') {
      return question.answer
        ? moment(question.answer).format('MMMM Do YYYY, h:mm:ss a')
        : null
    }
    return question.answer;
  };

  render() {
    const { results } = this.state;
    if (!results) {
      return <div style={{ width: '100%', height: '100%', display: 'flex', justifyContent: 'center', paddingTop: 200 }}>
        <Icon type="loading" style={{ fontSize: 70, marginTop: 8 }} />
      </div>
    }
    return (
      <div style={{ margin: 10 }}>
        <h1>Results</h1>
        <Button type="primary" onClick={this.export}>Download Results</Button>
        <br />
        <br />
        {results.length
          ? <Collapse accordion>
            {results.map((result, index) => {
              const content = result.data.map(question => {
                return (
                  <span>
                    <strong>{question.prompt}: </strong> {this.formatAnswer(question)}
                    <br />
                  </span>
                )
              });
              return (
                <Panel header={`Result Received at ${moment(result.submissionTime).format('MMMM Do YYYY, h:mm:ss a')}`} key={index}>
                  <p>{content}</p>
                </Panel>
              )
            })}
          </Collapse>
          : <strong>No Results to View</strong>
        }
      </div>
    )
  }
}

export default ViewResults;