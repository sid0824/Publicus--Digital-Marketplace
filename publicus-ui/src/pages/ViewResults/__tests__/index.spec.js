import toJson from 'enzyme-to-json';
import React from "react";
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {API} from "aws-amplify";
import Results from '../../../fixtures/Results'
import FormConfig from '../../../fixtures/FormConfig'
import ViewResults from "../index";
import download from 'js-file-download';

configure({ adapter: new Adapter() });

jest.mock('js-file-download');

describe('ViewResults', () => {

  beforeEach(() => {
    API.get = jest.fn().mockImplementation((name, path, options) => {
      return path.includes('results')
        ? Results
        : { form: escape(JSON.stringify(FormConfig)) }
    })
  });

  afterEach(() => {

  });

  test('should render view results page', (done) => {
    const wrapper = shallow(
      <ViewResults
        location={{
          search: 'SURVEY_ID'
        }}
      />
    );
    setTimeout(() => {
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.state()).toMatchSnapshot();
      done();
    });
  });

  test('should download results as csv', (done) => {
    const wrapper = shallow(
      <ViewResults
        location={{
          search: 'SURVEY_ID'
        }}
      />
    );
    setTimeout(() => {
      wrapper.instance().export();
      expect(download.mock.calls[0]).toMatchSnapshot();
      done();
    });
  });

});