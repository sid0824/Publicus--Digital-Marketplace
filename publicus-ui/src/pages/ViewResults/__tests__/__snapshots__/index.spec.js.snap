// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewResults should download results as csv 1`] = `
Array [
  "\\"submissionTime\\",\\"r email\\",\\"What is Your... ?\\",\\"What time is the...?\\",\\"What date is the...?\\"
\\"2019-10-24T02:45:29.000Z\\",\\"a@b.com\\",\\"2\\",\\"2019-10-24T02:46:09.000Z\\",
\\"2019-10-24T02:46:25.000Z\\",\\"a@b.com\\",\\"2\\",\\"2019-10-24T02:46:09.000Z\\",",
  "resultsData.csv",
]
`;

exports[`ViewResults should render view results page 1`] = `
<div
  style={
    Object {
      "margin": 10,
    }
  }
>
  <h1>
    Results
  </h1>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[Function]}
    type="primary"
  >
    Download Results
  </Button>
  <br />
  <br />
  <Collapse
    accordion={true}
    bordered={true}
    expandIconPosition="left"
    openAnimation={
      Object {
        "appear": [Function],
        "enter": [Function],
        "leave": [Function],
      }
    }
  >
    <CollapsePanel
      header="Result Received at October 24th 2019, 12:45:29 pm"
      key="0"
    >
      <p>
        <span>
          <strong>
            r email
            : 
          </strong>
           
          a@b.com
          <br />
        </span>
        <span>
          <strong>
            What is Your... ?
            : 
          </strong>
           
          <br />
        </span>
        <span>
          <strong>
            What is Your... ?
            : 
          </strong>
           
          1
          <br />
        </span>
        <span>
          <strong>
            What is Your... ?
            : 
          </strong>
           
          false
          <br />
        </span>
        <span>
          <strong>
            What time is the...?
            : 
          </strong>
           
          12:46:09 pm
          <br />
        </span>
        <span>
          <strong>
            What date is the...?
            : 
          </strong>
           
          <br />
        </span>
        <span>
          <strong>
            What is Your... ?
            : 
          </strong>
           
          2
          <br />
        </span>
      </p>
    </CollapsePanel>
    <CollapsePanel
      header="Result Received at October 24th 2019, 12:46:25 pm"
      key="1"
    >
      <p>
        <span>
          <strong>
            r email
            : 
          </strong>
           
          a@b.com
          <br />
        </span>
        <span>
          <strong>
            What is Your... ?
            : 
          </strong>
           
          <br />
        </span>
        <span>
          <strong>
            What is Your... ?
            : 
          </strong>
           
          1
          <br />
        </span>
        <span>
          <strong>
            What is Your... ?
            : 
          </strong>
           
          true
          <br />
        </span>
        <span>
          <strong>
            What time is the...?
            : 
          </strong>
           
          12:46:09 pm
          <br />
        </span>
        <span>
          <strong>
            What date is the...?
            : 
          </strong>
           
          <br />
        </span>
        <span>
          <strong>
            What is Your... ?
            : 
          </strong>
           
          2
          <br />
        </span>
      </p>
    </CollapsePanel>
  </Collapse>
</div>
`;

exports[`ViewResults should render view results page 2`] = `
Object {
  "results": Array [
    Object {
      "data": Array [
        Object {
          "answer": "a@b.com",
          "prompt": "r email",
          "type": "textInput",
        },
        Object {
          "answer": null,
          "prompt": "What is Your... ?",
          "type": "numberInput",
        },
        Object {
          "answer": "1",
          "prompt": "What is Your... ?",
          "type": "select",
        },
        Object {
          "answer": null,
          "prompt": "What is Your... ?",
          "type": "switch",
        },
        Object {
          "answer": "2019-10-24T02:46:09.000Z",
          "prompt": "What time is the...?",
          "type": "timePicker",
        },
        Object {
          "answer": null,
          "prompt": "What date is the...?",
          "type": "datePicker",
        },
        Object {
          "answer": "2",
          "prompt": "What is Your... ?",
          "type": "radio",
        },
      ],
      "submissionTime": "2019-10-24T02:45:29.000Z",
    },
    Object {
      "data": Array [
        Object {
          "answer": "a@b.com",
          "prompt": "r email",
          "type": "textInput",
        },
        Object {
          "answer": null,
          "prompt": "What is Your... ?",
          "type": "numberInput",
        },
        Object {
          "answer": "1",
          "prompt": "What is Your... ?",
          "type": "select",
        },
        Object {
          "answer": true,
          "prompt": "What is Your... ?",
          "type": "switch",
        },
        Object {
          "answer": "2019-10-24T02:46:09.000Z",
          "prompt": "What time is the...?",
          "type": "timePicker",
        },
        Object {
          "answer": null,
          "prompt": "What date is the...?",
          "type": "datePicker",
        },
        Object {
          "answer": "2",
          "prompt": "What is Your... ?",
          "type": "radio",
        },
      ],
      "submissionTime": "2019-10-24T02:46:25.000Z",
    },
  ],
}
`;
